src/Network/Http/Adapter/Stream.php

```diff
@@ -257,18 +257,36 @@ protected function _buildSslContext(Request $request, $options)
      */
     protected function _send(Request $request)
     {
+        $deadline = false;
+        if (isset($this->_contextOptions['timeout']) && $this->_contextOptions['timeout'] > 0) {
+            $deadline = time() + $this->_contextOptions['timeout'];
+        }
+
         $url = $request->url();
         $this->_open($url);
         $content = '';
+        $timedOut = false;
+
         while (!feof($this->_stream)) {
+            if ($deadline !== false) {
+                stream_set_timeout($this->_stream, max($deadline - time(), 1));
```

Comment:
To be honest, this line is not my idea :stuck_out_tongue: PEAR's `HTTP/Requet2` handles time-out in this way.